app/role_test.go

```diff
@@ -0,0 +1,217 @@
+		actualRole, err := th.App.GetRoleByName(testData.channelRole.Name)
+		require.Nil(t, err)
+		require.NotNil(t, actualRole)
+		require.Equal(t, testData.channelRole.Name, actualRole.Name)
+		require.Equal(t, testData.shouldHavePermission, slices.IncludesString(actualRole.Permissions, testData.permission.Id), "row: %+v", testData.truthTableRow)
+	})
+}
+
+// testPermissionInheritance tests 48 combinations of scheme, permission, role data.
+func testPermissionInheritance(t *testing.T, testCallback func(t *testing.T, th *TestHelper, testData permissionInheritanceTestData)) {
+	th := Setup(t).InitBasic()
+	defer th.TearDown()
+
+	th.App.SetLicense(model.NewTestLicense(""))
+	th.App.SetPhase2PermissionsMigrationStatus(true)
+
+	permissionsDefault := []string{
+		model.PERMISSION_MANAGE_CHANNEL_ROLES.Id,
+		model.PERMISSION_MANAGE_PUBLIC_CHANNEL_MEMBERS.Id,
+	}
+
+	// Defer resetting the system scheme permissions
+	systemSchemeRoles, err := th.App.GetRolesByNames([]string{
+		model.CHANNEL_GUEST_ROLE_ID,
+		model.CHANNEL_USER_ROLE_ID,
+		model.CHANNEL_ADMIN_ROLE_ID,
+	})
+	require.Nil(t, err)
+	require.Len(t, systemSchemeRoles, 3)
+
+	// defer resetting the system role permissions
+	for _, systemRole := range systemSchemeRoles {
+		defer th.App.PatchRole(systemRole, &model.RolePatch{
+			Permissions: &systemRole.Permissions,
+		})
+	}
+
+	// Make a channel scheme, clear its permissions
+	channelScheme, err := th.App.CreateScheme(&model.Scheme{
+		Name:        model.NewId(),
+		DisplayName: model.NewId(),
+		Scope:       model.SCHEME_SCOPE_CHANNEL,
+	})
+	require.Nil(t, err)
+	defer th.App.DeleteScheme(channelScheme.Id)
+
+	team := th.CreateTeam()
+	defer th.App.PermanentDeleteTeamId(team.Id)
+
+	// Make a channel
+	channel := th.CreateChannel(team)
+	defer th.App.PermanentDeleteChannel(channel)
+
+	// Set the channel scheme
+	channel.SchemeId = &channelScheme.Id
+	channel, err = th.App.UpdateChannelScheme(channel)
+	require.Nil(t, err)
+
+	// Get the truth table from CSV
+	file, e := os.Open("tests/channel-role-has-permission.csv")
+	require.Nil(t, e)
+	defer file.Close()
+
+	b, e := ioutil.ReadAll(file)
+	require.Nil(t, e)
+
+	r := csv.NewReader(strings.NewReader(string(b)))
+	records, e := r.ReadAll()
+	require.Nil(t, e)
+
+	test := func(higherScopedGuest, higherScopedUser, higherScopedAdmin string) {
+		for _, roleNameUnderTest := range []string{higherScopedGuest, higherScopedUser, higherScopedAdmin} {
+			for i, row := range records {
+				// skip csv header
+				if i == 0 {
+					continue
+				}
+
+				higherSchemeHasPermission, e := strconv.ParseBool(row[0])
+				require.Nil(t, e)
+
+				permissionIsModerated, e := strconv.ParseBool(row[1])
+				require.Nil(t, e)
+
+				channelSchemeHasPermission, e := strconv.ParseBool(row[2])
+				require.Nil(t, e)
+
+				channelRoleIsChannelAdmin, e := strconv.ParseBool(row[3])
+				require.Nil(t, e)
+
+				shouldHavePermission, e := strconv.ParseBool(row[4])
+				require.Nil(t, e)
```

Comment:
This is great - makes it a lot easier to understand whats going on - not sure if this was there last time i looked at this

Reply:
It wasn't there, I was using `p`, `q`, `r`, `s`, so understandably it was harder to follow. Glad it's clearer now.